### DISCUSSION TITLE
invariant and equivariant graph networks

### DISCUSSION THREAD

**Discussion by JoshuaMitton** [pyg-team/pytorch_geometric#2849] [Ideas]
Hi, is it possible to implement the graph network layers from invariant and equivariant graph networks https://arxiv.org/abs/1812.09902 in pytorch geometric?

Looking at the 15 bases from the paper there are 5 that update nodes and 10 that updates edges. Looking at the node updates, I think they are, 

1) `x_i = x_i`
2) `x_i = \sum x_j`
3) `x_i = \sum e_ij`
4) `x_i = \sum e_ji`
5) `x_i = \sum e_~i`

In the message passing framework 1,2,3,4 are straightforward, but I don't see how to achieve 5. For 5 I am trying to say that node i would be updated by the edges that do not connect to node i. Is it possible to access the edges in the graph that do not connect to node i in the message function that is updating node i in pytorch geometric? 

Thanks,
Josh

**Comment by rusty1s** [comment]
I think the best way to implement that is via `negative_sampling`, that is, during forward propagation, you sample a subset of negative edges for propagation. IMO, using all negative edges will result in a dense adjacency matrix which is computational expensive. Nonetheless, you can get the full negative adjacency matrix via `to_dense_adj`, inverting the connections, and using `dense_to_sparse` afterwards.

**Reply by JoshuaMitton** [reply]
Ah sorry I don't think I was very clear with my question. For the 5th one where I said edges that don't connect to node i, i.e. the current node being updated, but what I meant was say nodes i is node 1 then I would want to update node i using edge values between nodes 2-3 and nodes 2-4 and 3-4, etc. 

Is there a way to access these edges in the message passing framework? 

Thanks,
Josh

**Reply by rusty1s** [reply]
I'm sorry for misunderstanding your question, but I'm still not sure I understand it yet. What do you mean by edge values between nodes 2-3? In theory, you can get that values via `mask = (edge_index[0] == 2) & (edge_index[1] == 3)`. Is that what you mean?

**Reply by JoshuaMitton** [reply]
Yes, I suppose that, but I was wondering if there is a specific way to access that in the message passing framework, more specifically in the message function. So, if you image a graph with 3 nodes and numbers them 1,2,3 for the sake of explaining. Then you could also say you have edges 12, 21, 13, 31, 23, 32 connecting them. In the message passing framework in the message function when I am updating node 1's features I can have access to x_i, i.e. node 1's features, x_j, i.e. node 2 or 3's features, and edge_attr, which would be the edge features for the egde connecting nodes i and j. So in this When I am updating node 1's features I can see node 1's features, I can also access node 2 and 3's features, and I can access edge features of edges 12, 21, 13, 31, but I cannot access edge features of edges 23, 32. Basically what I would like to be able to do is also update nodes 1's features using the edge features from edges 23 and 32.

Do you know if there is a good way to do that in the message function and in such a way that it is more general than that small example?

Thanks,
Josh

**Reply by rusty1s** [reply]
The `MessagePassing` interface will only take care of propagating messages between edges that have you passed into it. In order to do what you described above, you need to manually modify the initial `edge_index` representation to include those edges. If you want to treat these additional edges differently, you may also want to separate those structures into two matrices, e.g., `edge_index_1` and `edge_index_2`, and use both separately in a `MessagePassing` module.

**Reply by JoshuaMitton** [reply]
I am still not sure I fully understand. In the example above I would like to update node 1's features using the edge that connects nodes 2 and 3. In the edge_index the connection between 2 and 3 already exists, its just that I don't know how to use that value to update node 1s features. 

In the documentation for the message function it says Constructs messages to node i in analogy to \phi for each edge in (j,i) \in \Epsilon ...

Since I want to update node 1's features with the edge features for the edge connecting nodes 2 and 3, then it is like updating node i's features with an edge that is not in (j,i). Is that possible? Also, how would I add this to the edge index as 2-3 is already in the edge index.

I don't know if I am miss-understanding your explanation.

Also taking another look at the code for the messing passing class. Perhaps it is not the message function that needs changing as that can compute a message for all the edges in the graph, but maybe I would need to change the aggregation function such that it can aggregate edge features into nodes that the edge doesn't connect. Do you think this is possible?

Thanks,
Josh

**Reply by rusty1s** [reply]
Ok, thanks for the clarification. You are right that this is not supported by default. The `MessagePassing` interface will only take care of aggregating incoming messages/from direct neighbors. In order to make this work, I think you need to operate on two `edge_index` representations. The first represents the regular graph, and is used to build message/edge representations, and the second is used to denote which edge/message will be aggregated to which node (it basically has a sparse matrix shape of `[num_edges, num_nodes]`). For example:
```python
x = ... # Node features
edge_index = ... # Adjacency representation (denoting sparse matrix of shape [num_nodes, num_nodes])

x_src = edge_index[0]
x_dst = edge_index[1]
# Compute a message/edge representation
edge_repr = MLP(torch.cat([x_src, x_dst], dim=-1))

hyperedge_index = ... # Incidence representation (denoting sparse matrix of shape [num_edges, num_nodes])
out = scatter(edge_repr[hyperedge_index[0]], hyperedge_index[1], dim=0, dim_size=num_nodes)
```
I hope this is useful to you.

**Reply by JoshuaMitton** [reply]
Thanks for you help. This sounds like a good way to implement it.
